file/src/test/scala/akka/stream/alpakka/file/scaladsl/LogRotatorSinkSpec.scala

```diff
@@ -0,0 +1,227 @@
+              fileName = path.toString
+              Some(path)
+            } else {
+              None
+            }
+          }
+      }
+      val completion = Source(testByteStrings).runWith(LogRotatorSink(testFunction))
+      val result = Await.result(completion, 3.seconds)
+
+      val (contents, sizes) = readUpFilesAndSizesThenClean(files)
+      sizes should contain theSameElementsAs Seq(6006L)
+      contents should contain theSameElementsAs Seq(TestLines.mkString(""))
+    }
+    "write lines to multiple files due to filesize" in assertAllStagesStopped {
+      val (testFunction, files) = fileLengthFunctionFactory()
+      val completion =
+        Source(testByteStrings).runWith(LogRotatorSink(testFunction))
+      val result = Await.result(completion, 3.seconds)
+
+      val (contents, sizes) = readUpFilesAndSizesThenClean(files())
+      sizes should contain theSameElementsAs Seq(2002L, 2002L, 2002L)
+      contents should contain theSameElementsAs TestLines.sliding(2, 2).map(_.mkString("")).toList
+    }
+    "upstream fail before first file creation" in assertAllStagesStopped {
+      val (testFunction, files) = fileLengthFunctionFactory()
+      val (probe, completion) =
+        TestSource.probe[ByteString].toMat(LogRotatorSink(testFunction))(Keep.both).run()
+
+      val ex = new Exception("my-exception")
+      probe.sendError(ex)
+      the[Exception] thrownBy Await.result(completion, 3.seconds) shouldBe ex
+      files() shouldBe empty
+    }
+
+    "upstream fail after first file creation" in assertAllStagesStopped {
+      val (testFunction, files) = fileLengthFunctionFactory()
+      val (probe, completion) =
+        TestSource.probe[ByteString].toMat(LogRotatorSink(testFunction))(Keep.both).run()
+
+      val ex = new Exception("my-exception")
+      probe.sendNext(ByteString("test"))
+      probe.sendError(ex)
+      the[Exception] thrownBy Await.result(completion, 3.seconds) shouldBe ex
+      files().size shouldBe 1
+      readUpFilesAndSizesThenClean(files())
+    }
+    "function fail on path creation" in assertAllStagesStopped {
+      val ex = new Exception("my-exception")
+      val testFunction = () => { (x: ByteString) =>
+        {
+          throw ex
+        }
+      }
+      val (probe, completion) =
+        TestSource.probe[ByteString].toMat(LogRotatorSink(testFunction))(Keep.both).run()
+      probe.sendNext(ByteString("test"))
+      the[Exception] thrownBy Await.result(completion, 3.seconds) shouldBe ex
+    }
+    "downstream fail on file write" in assertAllStagesStopped {
+      val path = Files.createTempFile(fs.getPath("/"), "test", ".log")
+      val testFunction = () => { (x: ByteString) =>
+        {
+          Option(path)
+        }
+      }
+      val (probe, completion) =
+        TestSource.probe[ByteString].toMat(LogRotatorSink(testFunction, Set(StandardOpenOption.READ)))(Keep.both).run()
+      probe.sendNext(ByteString("test"))
+      probe.expectCancellation()
```

Comment:
Yapp, its looks like that the cancle(in) (or the failstage) in the callback function not closes the upstream properly.